pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,444 @@
+		return nil, false, fmt.Errorf("Invalid IP inserted into IdentityMetadata: %s", prefix)
+	} else if ip.To4() != nil {
+		prefix = p + "/32"
+	} else {
+		prefix = p + "/128"
+	}
+
+	_, cidr, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, false, err
+	}
+
+	allLbls := cidrlabels.GetCIDRLabels(cidr)
+	allLbls.MergeLabels(lbls)
+
+	log.WithFields(logrus.Fields{
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the identityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+
+	var matching []string
+	sortedFilter := filter.SortedList()
+	for prefix, lbls := range identityMetadata {
+		if bytes.Contains(lbls.SortedList(), sortedFilter) {
+			matching = append(matching, prefix)
+		}
+	}
+	return matching
+}
+
+// RemoveLabelsFromIPs wraps RemoveLabels to provide a convenient method for
+// the caller to remove all given prefixes at once. This function will trigger
+// policy update and recalculation if necessary on behalf of the caller if any
+// changes to the kube-apiserver were detected.
+//
+// Identities allocated by InjectLabels() may be released by RemoveLabels().
+//
+// A prefix will only be removed from the IDMD if the set of labels becomes
+// empty.
+func RemoveLabelsFromIPs(
+	m map[string]labels.Labels,
+	src source.Source,
+	updater identityUpdater,
+	triggerer policyTriggerer,
+) {
+	var (
+		idsToAdd    = make(map[identity.NumericIdentity]labels.LabelArray)
+		idsToDelete = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range m {
+		id, exists := IPIdentityCache.LookupByIP(prefix)
+		if !exists {
+			continue
+		}
+		// Insert to propagate the updated set of labels after removal.
+		var la labels.LabelArray
+		if l := RemoveLabels(prefix, lbls, src); l != nil {
+			la = l.LabelArray()
+		}
+		idsToDelete[id.ID] = la
+		if len(lbls) > 0 {
+			// If for example kube-apiserver label is removed from
+			// a remote-node, then RemoveLabels() will return a
+			// non-empty set representing the new full set of
+			// labels to associate with the node. In order to
+			// propagate the new identity, we must emit a delete
+			// event for the old identity and then an add event for
+			// the new identity.
+			idsToAdd[id.ID] = la
+		}
+	}
+	if len(idsToDelete) > 0 {
+		var wg sync.WaitGroup
+		// SelectorCache.UpdateIdentities() asks for callers to avoid
+		// handing the same identity in both 'adds' and 'deletes'
+		// parameters here, so make two calls. These changes will not
+		// be propagated to the datapath until later.
+		updater.UpdateIdentities(nil, idsToDelete, &wg)
+		updater.UpdateIdentities(idsToAdd, nil, &wg)
```

Comment:
Looking back through my git stash, I think there's a corner case here that is maybe not currently possible, but could hit someone really weirdly in future:

`id` above this block is some identity allocated based on a series of labels, like `reserved:kube:apiserver,cidr:w.x.y.z/n,...`. `RemoveLabels` will potentially then subtract one or more of those labels out of the list. So `l` ends up with say only `cidr:w.x.y.z/n`. There's no guarantee that there is any identity associated with that remaining set of labels at this point. Least of which, the identity `id` does *not* correspond to the new `l` set of labels.

From a quick peruse of the code, I think that `SelectorCache.UpdateIdentities()` will warn in this case, but I can't tell what the behaviour would actually be. Maybe one case to consider is if you had the apiserver on a remote node in the cluster, then you removed the kube-apiserver from that node (eg shift it to the current node). What happens to the selectorcache in that case? Do we suddenly change the cached labels for this identity?

cc @christarazi